metals/src/main/scala/scala/meta/internal/metals/debug/DebugProvider.scala

```diff
@@ -924,4 +920,81 @@ object DebugProvider {
       extends Exception(
         "Build misconfiguration. No semanticdb can be found for you file, please check the doctor."
       )
+
+  sealed abstract class ClassSearch[A](debugProvider: DebugProvider)(implicit
+      ec: ExecutionContext
+  ) {
+    private val searchStarted = new ju.concurrent.atomic.AtomicBoolean(false)
+    private val searchPromise = Promise[Try[A]]()
+    protected def search(): Try[A]
+    protected def dapSessionParams(res: A): Future[DebugSessionParams]
+    def createDapSession(args: A): Future[DebugSession] =
+      dapSessionParams(args).flatMap(debugProvider.asSession(_))
+    def searchResult: Future[(Option[A], ClassSearch[A])] = {
+      if (searchStarted.compareAndSet(false, true)) {
```

Comment:
Is it possible that this will be invoked multiple times? Looks like we invoke searchResult on each separate search only once

Reply:
Yes, we do. It's not needed for this use case.